Task/change add to layout button [DHIS2-5348] [DHIS2-5336]

joakimia opened the pull request:
This PR Includes:

Requested changes to the "AddToLayout" button displayed in dialogs for unselected Dimensions in left pane:

**Styling changes:**
- "ADD TO SERIES" and "ArrowDown" button have color="primary", (Blue).
-  MenuItems in the DropDown Menu have been explicitly defined as a  list item component.

**Functionality changes:** 
- AddToLayoutButton now triggers an update by dispatching the action _setCurrentFromUi_

Additional info:
- Compared to the capture app, the dropDown icon is still an "ArrowDown" icon instead of an "ArrowUp".
- MenuItems and AddToLayout Button text are still rendered in ALL CAPS.

I will verify with Joe and show him the PR to see if we should use _text-transform_ such that button text / MenuItem text will be rendered with only a starting capital letter  ( "Add to series" instead of "ADD TO SERIES").

Also, i'll double check with Joe if we really want to change the DropDown Icon from "ArrowDown" to "ArrowUp" to mirror the capture app button.

![image](https://user-images.githubusercontent.com/13645152/48966882-e2180700-efd8-11e8-8f3f-7d8ca740122d.png)

joeDHIS commented:
Ideally the look of the dropdown menu should mirror a standard MUI/d2ui menu. So the required changes to this would be: left align the text, use Sentence case (not all caps), font: Roboto regular 15px. (See menu component here: https://d2-ci.github.io/ui/)
Don't worry about changing the icon up/down when the menu is open.

joakimia commented:
After addressing the comments @joeDHIS provided:

closed state:
![image](https://user-images.githubusercontent.com/13645152/49003353-0a7f3d00-f162-11e8-8173-532aeb0f9a94.png)


Opened state:
![image](https://user-images.githubusercontent.com/13645152/49003382-1834c280-f162-11e8-84f5-71ff6f01310d.png)

amcgee commented on packages/app/src/components/AddToLayoutButton/AddToLayoutButton.js:
```diff
@@ -47,35 +50,45 @@ export class AddToLayoutButton extends Component {
 
-    getMenuItems = () =>
-        items.slice(1).map(option => (
+    omitSeriesMenuItem = () =>
```
One minor detail - I like function names to be descriptive and lexically / grammatically correct.  So in this case `omit` is a verb, so I think it should be acting on an argument, so say `omitSeriesMenuItem(items)`.  If it instead acts on the `this` object I might make it a method rather than a function and implement `items.withoutSeriesMenuItem()` or something (note that I'm using the preposition `without` instead of the imperative verb conjugation `omit` here)

All this being said, I actually think this is superfluous to implement as a class method.  The result is derived only from a constant, not the state of the AddToLayoutButton class instance, and so should be a constant.  Instead of making this a class method I would just create the constant subset list above, with the other Consts (and maybe consider doing the same for the SERIES and FILTER indexes used below):

```js
const seriesItem = ADD_LAYOUT_OPTIONS[0];
const filterItem = ADD_LAYOUT_OPTIONS[2];
const itemsWithoutSeries = ADD_LAYOUT_OPTIONS.filter(option => option.axisKey !== 'columns');
/*OR, perhaps more descriptive? */ const nonColumnItems = ADD_LAYOUT_OPTIONS.filter(option => option.axisKey !== 'columns');
```

Then just use those const values in the body of the code, since their values will never change.

On another note - I think renaming `ADD_TO_LAYOUT_OPTIONS` to `items` is a little misleading, since it's a constant value while the name `items` implies a variable.  Similarly, I used variable-style names for my consts in this example but might consider something like FILTER_ITEM instead of filterItem (the sliced array is a bit of a gray area since it is the result of a computation on a constant, not a constant itself).

That was a bit long-winded, and this is a very small detail but I think it helps to have code be as readable and understandable as possible for maintainability, so I'm sharing a few of my opinions though I'm sure other people will disagree with some of them!

amcgee commented on packages/app/src/components/AddToLayoutButton/Menu.js:
```diff
@@ -1,40 +1,58 @@
+    const isOverFlowing =
+        anchorEl &&
+        anchorEl.getBoundingClientRect().bottom + MENU_HEIGHT >
+            window.innerHeight;
```
Discussed out-of-band, 100px is a good-enough default since it would be challenging to calculate this on the fly.

joeDHIS commented:
Austins points:
1) I think, on seeing this again, we should try to match the the menu to the width of the button if possible,
2) rounded corners are correct, we are using a menu component from MUI (and d2ui, eventually)
3) Should be fixed, ideally.

amcgee commented:
Thanks @joeDHIS.  @joakimia are you able to take a look at those points?

joakimia commented:
I've set minWidth on the DropDown to align itself with the Add to series button.
The width of the "Add to series" button might increase though in other languages, depending on the length of the string, but the same goes for the Menu Items.

I've also increased the threshold for rendering the dropdown upwards by 20 pixels.

@amcgee  take a new look?
